Get enterprise details

Jakeamorris25 opened the pull request.

Jakeamorris25 committed:
Added functionality to Get(id) method

Still need to add password authorisation.

Jakeamorris25 committed:
Added password protection / Unit tests

Jakeamorris25 committed:
Cleaned up files

danielelkington commented on Backend/Backend.WebServices/Controllers/EnterpriseController.cs:
```diff
@@ -51,11 +51,26 @@ public IEnumerable<EnterpriseDTO> Get()
-        public string Get(int id)
+        /// <param name="password">Password of the enterprise</param>
+        /// <returns>An EnterpriseDTO with the specified ID</returns>
+        public EnterpriseDTO Get(int id, string password = null)
```
The password should not be a parameter in the GET request, instead we should get it from the headers. Can do this in the body of the method, something like 
```
string password = null;
if (Request.Headers.Any(x => "Authorization".Equals(x.Key)))
{
    password = Request.Headers.First(x => "Authorization".Equals(x.Key));
}
```

danielelkington commented on Backend/Backend.WebServices/Controllers/EnterpriseController.cs:
```diff
@@ -51,11 +51,26 @@ public IEnumerable<EnterpriseDTO> Get()
         {
             ThrowExceptionIfNotHttps();
-            return "ForestedgeCommunityGarden";
+            var enterprises = _context.Enterprises.ToList();
```
It's probably not a good idea to do a ToList() on the whole enterprises context - this will immediately fire off a request to the database that gets all the enterprises, when for performance reasons we should only get one.
Also, this will only include the items in the Enterprise table, when we want to do joins to the Participant, Place, and Media Items in the same query (using the db foreign keys).
The actual query to get the single enterprise should therefore be something like:
```
var enterprise = _context.Enterprises
    .Where(x => x.Id == id)
    .Include(x => x.Participants.Select(p => p.Places.Select(pl => pl.MediaItems.Select(m => m.MediaItemType))))
    .SingleOrDefault();
if (enterprise == null)
{
    //throw appropriate exception - the id they've sent us isn't a valid enterprise ID!
}
```
All the stuff in the include line will be returned with the enterprise (ie the enterprise will have all its participants, places, media items and media item types populated in the object graph.

danielelkington commented on Backend/Backend.WebServices/Controllers/EnterpriseController.cs:
```diff
@@ -51,11 +51,26 @@ public IEnumerable<EnterpriseDTO> Get()
+            if (enterprise == null)
+                throw new HttpException(404, "Enterprise not found");
+
+            return new EnterpriseDTO(enterprise);
```
The EnterpriseDTO is not a good DTO to use here; it includes a very slimmed-down version of the enterprise that is only suitable for showing a list of enterprises. Instead we want to return a new DTO that includes all the details of the enterprise, including participants, places, etc.

I'd suggest making a class called EnterpriseDetailDTO, along with the classes ParticipantDTO, PlaceDTO, etc. You'd then need to write some code in each of these to map the Database classes to the DTOs.

danielelkington commented on Backend/Backend.UnitTest/EnterpriseControllerTest.cs:
```diff
@@ -62,8 +62,14 @@ public void GetEnterprises_ReturnsEnterprisesInAlphabeticalOrderByName()
         [TestMethod]
         public void GetEnterprise_ReturnsSingleEnterprise()
```
See the changes below - once you make them you may want to consider writing some more tests, with things like making sure the correct exception is thrown if the enterprise ID doesn't exist or if the password is wrong, and making sure that all the enterprise details (including participants, places, etc) get mapped correctly in the object graph that gets returned.

danielelkington reviewed (changes requested):
Excellent start considering this is your first time with ASP.net and Entity Framework! Let me know if any of the changes requested below don't make sense.

Jakeamorris25 committed:
Added additional DTOs / unit tests and improved EnterpriseController.

Jakeamorris25 committed:
Added PlaceDTO

danielelkington commented on Backend/Backend.UnitTest/EnterpriseControllerTest.cs:
```diff
@@ -62,14 +65,70 @@ public void GetEnterprises_ReturnsEnterprisesInAlphabeticalOrderByName()
+            try
+            {
+                var response = _target.Get(5);
+            }
```
When I do tests like this I usually put Assert.Fail() at the end of the try to make sure that an exception really is thrown!

danielelkington commented on Backend/Backend.WebServices/Controllers/EnterpriseController.cs:
```diff
@@ -53,24 +54,31 @@ public IEnumerable<EnterpriseDTO> Get()
+            {
+                password = Request.Headers.GetValues("Authorization").First();
+
+                if (!password.Equals(enterprise.Password))
```
Do we care if the client sends a password for an enterprise that doesn't have one? What if they sent an Authorization header with an empty value in cases where the enterprise doesn't have a password? (Then again, maybe I'm just being nitpicky :/ )

danielelkington commented on Backend/Backend.WebServices/DataTransferObjects/EnterpriseDetailsDTO.cs:
```diff
@@ -0,0 +1,29 @@
+            CoverImage = enterprise.CoverImageURL;
+            CoverCoordinate = enterprise.CoverCoordinate;
+            ModifiedUTC = enterprise.ModifiedUTC;
+            Participants = enterprise.Participants;
```
I think the Participants list should be of type ParticipantDTO, and this line should be something like
`Participants = enterprise.Participants.Select(x => new ParticipantDTO(x)).ToList()`
And the other DTOs should do similar things for the child entities. Otherwise it will return an EnterpriseDTO with a list of Participant database objects :/

danielelkington reviewed (commented):
Almost there, just a couple more points (most important one is mapping the DTOs going back)

Jakeamorris25 committed:
Converted entities to their DTOs

Also stopped checking for a password if the enterprise isnt password
protected.

danielelkington commented on Backend/Backend.WebServices/DataTransferObjects/MediaItemTypeDTO.cs:
```diff
@@ -9,11 +10,11 @@ public MediaItemTypeDTO(MediaItemType mediaItemType)
         public int Id { get; set; }
         public string Name { get; set; }
-        public virtual ICollection<MediaItem> MediaItems { get; set; }
+        public virtual ICollection<MediaItemDTO> MediaItems { get; set; }
```
Hmm...should our MediaItemTypeDTO have a link back to MediaItems? So our object graph will be Enterprise -> Participant -> Place -> MediaItem -> MediaItemType -> MediaItem -> MediaItemType -> MediaItem...

The database object needs to have the relationship going both ways because of the foreign key, but our DTOs shouldn't have this two-way relationship. I think we can remove the MediaItems list from the MediaItemTypeDTO class.

danielelkington reviewed (approved):
I think literally two lines need to be removed, and it's all good to merge!
Edit: I've made the changes (and a few others). It's deployed to Azure now and seems to work well.

Commit:
Get Enterprise Details enhancements
-Improving DTOs to reflect data returned
-Fixing a couple of bugs, including one requiring a certain assembly to be loaded in order to use dbgeography

danielelkington committed:
Merge branch 'master' into get-enterprise-details
